### PR TITLE
Fixed 8 important bugs and two dozen minor bugs from findbugs

### PR DESCRIPTION
**Pull request by DrDub**
The details of the bugs fixed are in the commit log.

Findbugs was executed in the sphinx4-core folder after a successful gradle build as follows:

sphinx4-core$ findbugs -textui -auxclasspath ~/.m2/repository/org/apache/commons/commons-math3/3.2/commons-math3-3.2.jar  -sortByClass -high build/classes/main

There are still close to 100 "relying in default encoding" bugs. That might affect people running sphinx4 in environments where UTF-8 is not the default encoding.

### CODE REVIEW

File: sphinx4-core/src/main/java/edu/cmu/sphinx/util/props/ConfigurationManagerUtils.java [hunk 1]
```diff
@@ -660,10 +663,10 @@ public static void setProperty(ConfigurationManager cm, String componentName, St
                 ps.setBoolean(propName, Boolean.valueOf(propValue));
                 break;
             case DOUBLE:
-                ps.setDouble(propName, new Double(propValue));
```
Comment (mbait) [inline sphinx4-core/src/main/java/edu/cmu/sphinx/util/props/ConfigurationManagerUtils.java]: Can you explain this change? `new Double` creates an object and `Double#parseDouble()` a primitive which is preferable.

File: sphinx4-core/src/main/java/edu/cmu/sphinx/alignment/tokenizer/DecisionTree.java [hunk 2]
```diff
@@ -182,19 +182,19 @@ public void dumpDot(PrintWriter out) {
         out.write("rankdir = LR\n");
 
         for (Node n : cart) {
-            out.println("\tnode" + Math.abs(n.hashCode()) + " [ label=\""
+            out.println("\tnode" + (n.hashCode() & Integer.MAX_VALUE) + " [ label=\""
```
Comment (mbait) [inline sphinx4-core/src/main/java/edu/cmu/sphinx/alignment/tokenizer/DecisionTree.java]: Previous version was clear - we take absolute of the hash code. The new version is not clear at all.

File: sphinx4-core/src/main/java/edu/cmu/sphinx/alignment/tokenizer/DecisionTree.java [hunk 3]
```diff
@@ -182,19 +182,19 @@ public void dumpDot(PrintWriter out) {
...
                     + n.toString() + "\", color=" + dumpDotNodeColor(n)
                     + ", shape=" + dumpDotNodeShape(n) + " ]\n");
             if (n instanceof DecisionNode) {
                 DecisionNode dn = (DecisionNode) n;
                 if (dn.qtrue < cart.length && cart[dn.qtrue] != null) {
-                    out.write("\tnode" + Math.abs(n.hashCode()) + " -> node"
-                            + Math.abs(cart[dn.qtrue].hashCode())
+                    out.write("\tnode" + (n.hashCode() & Integer.MAX_VALUE) + " -> node"
```
Comment (mbait) [inline sphinx4-core/src/main/java/edu/cmu/sphinx/alignment/tokenizer/DecisionTree.java]: Same as above.

File: sphinx4-core/src/main/java/edu/cmu/sphinx/alignment/tokenizer/DecisionTree.java [hunk 4]
```diff
@@ -182,19 +182,19 @@ public void dumpDot(PrintWriter out) {
...
+                            + (cart[dn.qtrue].hashCode() & Integer.MAX_VALUE)
```
Comment (mbait) [inline sphinx4-core/src/main/java/edu/cmu/sphinx/alignment/tokenizer/DecisionTree.java]: Same.

File: sphinx4-core/src/main/java/edu/cmu/sphinx/alignment/tokenizer/DecisionTree.java [hunk 5]
```diff
@@ -182,19 +182,19 @@ public void dumpDot(PrintWriter out) {
...
                             + " [ label=" + "TRUE" + " ]\n");
                 }
                 if (dn.qfalse < cart.length && cart[dn.qfalse] != null) {
-                    out.write("\tnode" + Math.abs(n.hashCode()) + " -> node"
-                            + Math.abs(cart[dn.qfalse].hashCode())
+                    out.write("\tnode" + (n.hashCode() & Integer.MAX_VALUE) + " -> node"
```
Comment (mbait) [inline sphinx4-core/src/main/java/edu/cmu/sphinx/alignment/tokenizer/DecisionTree.java]: Same.

File: sphinx4-core/src/main/java/edu/cmu/sphinx/decoder/search/WordPruningBreadthFirstSearchManager.java [hunk 6]
```diff
@@ -146,7 +146,7 @@
     protected long currentCollectTime; // the current frame number
     protected ActiveList activeList; // the list of active tokens
     protected List<Token> resultList; // the current set of results
-    protected Map<SearchState, Token> bestTokenMap;
+    protected Map<SearchState, Object> bestTokenMap;
```
Comment (mbait) [inline sphinx4-core/src/main/java/edu/cmu/sphinx/decoder/search/WordPruningBreadthFirstSearchManager.java]: Why is this required?

File: sphinx4-core/src/main/java/edu/cmu/sphinx/fst/Fst.java [hunk 7]
```diff
@@ -401,13 +401,26 @@ public boolean equals(Object obj) {
     /*
      * (non-Javadoc)
      * 
+     * @see java.lang.Object#hashCode()
+     */
+    @Override
+    public int hashCode() {
+        return Arrays.hashCode(isyms) + Arrays.hashCode(osyms) +
```
Comment (mbait) [inline sphinx4-core/src/main/java/edu/cmu/sphinx/fst/Fst.java]: Previously it was advised to use polynomial hash algorithm as the default one, but my knowledge may be outdated here.

File: sphinx4-core/src/main/java/edu/cmu/sphinx/fst/operations/Determinize.java [hunk 8]
```diff
@@ -118,7 +118,7 @@ public static Fst get(Fst fst) {
         while (!queue.isEmpty()) {
             ArrayList<Pair<State, Float>> p = queue.remove();
             State pnew = getStateLabel(p, stateMapper);
-            queue.remove(0);
+            //queue.remove(0);
```
Comment (mbait) [inline sphinx4-core/src/main/java/edu/cmu/sphinx/fst/operations/Determinize.java]: Em... any comments about that?

**Comment by mbait** [comment]
I'd say the the whole diff is not easy to review, some moments are unclear. I'd rather split this diff into parts each of which fixes a single bug, so you could explain the rationale behind the change.

**Comment by mbait** [comment]
Oh, I just realized that you made it into the commit message. I'm very sorry, will read it now.

**Comment by DrDub** [comment]
I just followed findbugs comments. Feel free to reject any changes you feel don't make the code easier / understandable. There are a few bugs that are real bugs and it would be nice to see them fixed thoug. I'll be glad to go through them one by one (I had to do some reading to understand what was the bug findbugs was complaining about, I can share the links I got), just add another ping to your earlier comments and we can discuss. I can then "water down" the PR by making fewer changes.

**Comment by mbait** [comment]
Thanks for the work, it's highly appreciated! I accepted and committed the most of changes. Unfortunately, I couldn't use automatic merge as the main work is still being done on SourceForge, so the only credit I could give to you was mention in the commit message.
